File: powerfulsharedpreferences/src/main/java/com/stefanosiano/powerfulsharedpreferences/Preferences.java
package com.stefanosiano.powerfulsharedpreferences;

/**
 * PowerfulPreference wrapper for Double
 */

class BPreference extends PowerfulPreference<Boolean> {
    BPreference(String key, Boolean defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public Boolean parse(String s) {return Boolean.parseBoolean(s);}
    @Override public Class getPrefClass() {return Boolean.class;}
}
class IPreference extends PowerfulPreference<Integer> {
    IPreference(String key, Integer defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public Integer parse(String s) {return Integer.parseInt(s);}
    @Override public Class getPrefClass() {return Integer.class;}
}
class LPreference extends PowerfulPreference<Long> {
    LPreference(String key, Long defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public Long parse(String s) {return Long.parseLong(s);}
    @Override public Class getPrefClass() {return Long.class;}
}
class FPreference extends PowerfulPreference<Float> {
    FPreference(String key, Float defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public Float parse(String s) {return Float.parseFloat(s);}
    @Override public Class getPrefClass() {return Float.class;}
}
class DPreference extends PowerfulPreference<Double> {
    DPreference(String key, Double defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public Double parse(String s) {return Double.parseDouble(s);}
    @Override public Class getPrefClass() {return Double.class;}
}
class SPreference extends PowerfulPreference<String> {
    SPreference(String key, String defaultValue, String prefName) {super(key, defaultValue, prefName);}
    @Override public String parse(String s) {return s;}
    @Override public Class getPrefClass() {return String.class;}
}

File: powerfulsharedpreferences/src/main/java/com/stefanosiano/powerfulsharedpreferences/Preferences.kt
package com.stefanosiano.powerfulsharedpreferences


internal class BPreference(key: String, defaultValue: Boolean, prefName: String?) : PowerfulPreference<Boolean>(key, defaultValue, prefName) {
    override fun getPrefClass() = Boolean::class.java
    override fun parse(s: String) = s.toBoolean()
    override fun toPreferences(value: Boolean): String = value.toString()
}

internal class IPreference(key: String, defaultValue: Int, prefName: String?) : PowerfulPreference<Int>(key, defaultValue, prefName) {
    override fun getPrefClass() = Int::class.java
    override fun parse(s: String) = s.toInt()
    override fun toPreferences(value: Int): String = value.toString()
}

internal class LPreference(key: String, defaultValue: Long, prefName: String?) : PowerfulPreference<Long>(key, defaultValue, prefName) {
    override fun getPrefClass() = Long::class.java
    override fun parse(s: String) = s.toLong()
    override fun toPreferences(value: Long): String = value.toString()
}

internal class FPreference(key: String, defaultValue: Float, prefName: String?) : PowerfulPreference<Float>(key, defaultValue, prefName) {
    override fun getPrefClass() = Float::class.java
    override fun parse(s: String) = s.toFloat()
    override fun toPreferences(value: Float): String = value.toString()
}

internal class DPreference(key: String, defaultValue: Double, prefName: String?) : PowerfulPreference<Double>(key, defaultValue, prefName) {
    override fun getPrefClass() = Double::class.java
    override fun parse(s: String) = s.toDouble()
    override fun toPreferences(value: Double): String = value.toString()
}

internal class SPreference(key: String, defaultValue: String, prefName: String?) : PowerfulPreference<String>(key, defaultValue, prefName) {
    override fun getPrefClass() = String::class.java
    override fun parse(s: String) = s
    override fun toPreferences(value: String): String = value
}

internal class DummyPreference(key: String, defaultValue: Any?, prefName: String?) : PowerfulPreference<String>(key, defaultValue?.toString()?:"", prefName) {
    override fun getPrefClass() = String::class.java
    override fun parse(s: String) = s
    override fun toPreferences(value: String): String = value
}

internal class EnumPreference<T>(private val clazz: Class<T>, key: String, defaultValue: T, prefName: String?) : PowerfulPreference<T>(key, defaultValue, prefName) where T: Enum<T> {
    override fun getPrefClass() = clazz
    override fun parse(s: String): T = clazz.enumConstants.first { it.name == s }
    override fun toPreferences(value: T): String = value.name
}